crengine/src/lvtextfm.cpp

```diff
@@ -1901,9 +1901,9 @@ class LVFormatter {
                         if ( lastSrc->flags & LTEXT_FLAG_PREFORMATTED )
                             scale_space_width = false;
                     }
-                    int cumulative_width_removed = 0;
-                    int prev_orig_measured_width = 0;
-                    int char_width = 0; // current single char width
+                    lUInt16 cumulative_width_removed = 0;
+                    lUInt16 prev_orig_measured_width = 0;
+                    lUInt16 char_width = 0; // current single char width
```

Comment:
Does this fixes something ? Or you just made them lUInt16 because you know you were working with lUInt16 as returned by lvfntman's measureText() ?
It shouldn't hurt keeping them int, right ? As a int can't be overflowed when we give it lUInt16 ?
But I think `cumulative_width_removed` can get a chance to be negative (at line 1922), so better it stays `int` (or all stays `int` to take the less risks :)

Reply:
To be honest, I am not ready to investigate all of this, but if you are sure that this change is doing more harm, I am ready to revert it.

Reply:
I'm not sure it does more harm, but it feels it can :)
I would let lvtextvm.cpp unchanged.

> But I mean it is not the size that is important here, but the sign

Yes, but if we get lUInt16, we can easily store/add them to a `int` - the `int` will stay positive, and shouldn't hurt if we put/add it back in a lUInt16 (for individual ones) or accumulate them (in which case a `unsigned int` would allow more, but we're fine with just a `int` if we don't want to spend much time thinking about it and possible side effects).

But I think your other changes in lvfreetypeface.cpp are right (hard to get back into all that :).
Btw, can you share the document from your screenshot (or part of it)? I didn't manage to reproduce the bug that this should solve, even by making a super long paragraph in a fb2 - might depends on the font setups, dunno...

Reply:
I think paragraph with 2000 characters long will be enough with font size about 80 px.
Set main font without characters from main book language (I use "Android Clock"), set HarfBuzz render and try. On my side the first fallback font is also not suitable ("Noto Color Emoji"), but I think it has no effect.

[Chadovich_Tropa_1_Evangelie-ot-Timofeya.vNFIxg.167219.fb2.zip](https://github.com/buggins/coolreader/files/7076209/Chadovich_Tropa_1_Evangelie-ot-Timofeya.vNFIxg.167219.fb2.zip)
( Scroll to last page )

Reply:
Thanks, I could reproduce.
I think your fix is fine - but I'm not really familiar with the lastFit handling.

(There is this:
```diff
             // i is used below to "fill props for rest of chars", so make it accurate
-            i = len; // actually make it do nothing
+            i = lastFitChar;
```
My memory sucks, and I don't remember if I had a good reason for this "actually make it do nothing" or if I was just too lazy understand why filling flags for next ignored chars was needed - but I also have the vague memory it was needed in some other context, but dunno if that was before Harfbuzz cluster rewrite or after...
So, let's just keep that in mind in case we notice some issue later.
)

Reply:
> but I'm not really familiar with the lastFit handling.

I don’t know, this is mostly your code. But after analyzing it, it seemed to me that you simply did not leave unprocessed characters, so that's why. Now this is no longer the case and I did it by analogy with a simple FreeType render.